### DISCUSSION TITLE
Single table 'closest' schema logic question

### DISCUSSION THREAD

**Discussion by mtt88** [dynamoose/dynamoose#1204] [Q&A] [answered]
Hi, I was wondering if you could provide a little more detail (or point me in the direction of the code) relating to how Dynamoose will select the 'closest' schema when working on a single table with multiple associated schemas.

Specifically the section of [documentation](https://dynamoosejs.com/guide/Model/)

> Behind the scenes Dynamoose will automatically pick the closest schema to match to your document, and use that schema for all operations pertaining to that document. If no matching schema can be found, it will default to the first schema in the array.

What logic or rules will determine 'closest' schema? 

Thanks

**Comment by fishcharlie** [comment, marked as answer]
@mtt88 Very very high level without diving into the code:

For each schema, it will combine that schemas attributes with the attributes for your document. Each attribute will get a score, 1 being full match, and 0 being an error (invalid type, etc.). There is also a value for 0.5 which represents that the attribute exists in either your document or schema but doesn't exist in the other. Once each attribute has a value, the value for that entire schema will be the lowest attribute value (worst attribute for that schema).

Now that we have a value for each schema, the highest schema value will get chosen (representing the best schema).

The `Constant` type in Dynamoose helps with this a lot because it will cause that attribute to be 0 if invalid, which helps promote and demote certain schemas. This is especially helpful if you have two schemas that are basically identical (which is a rare but possible edge case, if you are setting default values or doing other advanced modifying behaviors).

---

I have attached a few snippets from the Dynamoose codebase that dive into this a bit deeper:

https://github.com/dynamoose/dynamoose/blob/40443acc6f7a07681592f294d3893902c3652e34/lib/Model/index.ts#L418-L424

https://github.com/dynamoose/dynamoose/blob/40443acc6f7a07681592f294d3893902c3652e34/lib/Schema.ts#L405-L474

---

Hopefully that answers your question. If you have any suggestions or use cases that don't fit this model, I'd love to hear it. But I think it covers everything fairly well.

**Reply by mtt88** [reply]
Thanks for your quick reply - this was exactly the detail I was looking for.

My question was mostly to understand the 'cost' associated with deriving the closest schema and whether this might be something to take into consideration when deciding to associate multiple schemas with a single table/model.

Matt

**Reply by fishcharlie** [reply]
@mtt88 I think single table design if done properly is the best in terms of performance cost for DynamoDB. Tho I haven't done any specific performance tests on the code I showed above.

Hopefully at some point I'll get around to adding the ability to manually select which schema is used. But a lot of this behavior is changing in v3, so that might not even be necessary anymore.

Glad to have helped tho!
